From 700c94a79ed1ccabf32081c86fe282af4747e5e8 Mon Sep 17 00:00:00 2001
From: Chen Wandun <chenwandun@huawei.com>
Date: Wed, 9 Feb 2022 15:36:32 +0800
Subject: [PATCH 3535/4585] [Huawei] proc/meminfo: Add "FileCache" item in
 /proc/meminfo

Offering: HULK
hulk inclusion
category: feature
bugzilla: 187270

--------------------------------

Item "FileCache" in /proc/meminfo show the number of page cache
in LRU(active + inactive).

Signed-off-by: Chen Wandun <chenwandun@huawei.com>
---
 include/linux/mem_reliable.h | 12 ++++++++++++
 mm/mem_reliable.c            |  8 ++++++++
 2 files changed, 20 insertions(+)

diff --git a/include/linux/mem_reliable.h b/include/linux/mem_reliable.h
index 89b00799deb6..d8ddd03de724 100644
--- a/include/linux/mem_reliable.h
+++ b/include/linux/mem_reliable.h
@@ -56,6 +56,17 @@ static inline bool shmem_reliable_is_enabled(void)
 {
 	return shmem_reliable;
 }
+
+static inline bool page_reliable(struct page *page)
+{
+	if (!mem_reliable_is_enabled())
+		return false;
+
+	if (!page)
+		return false;
+
+	return page_zonenum(page) < ZONE_MOVABLE;
+}
 #else
 #define reliable_enabled 0
 #define pagecache_use_reliable_mem 0
@@ -74,6 +85,7 @@ static inline void reliable_report_meminfo(struct seq_file *m) {}
 static inline bool shmem_reliable_is_enabled(void) { return false; }
 static inline void page_cache_prepare_alloc(gfp_t *gfp) {}
 static inline bool mem_reliable_status(void) { return false; }
+static inline bool page_reliable(struct page *page) { return false; }
 #endif
 
 #endif
diff --git a/mm/mem_reliable.c b/mm/mem_reliable.c
index 32b277ddcd0b..faf7c0dbeda3 100644
--- a/mm/mem_reliable.c
+++ b/mm/mem_reliable.c
@@ -126,6 +126,14 @@ void reliable_report_meminfo(struct seq_file *m)
 	show_val_kb(m, "ReliableTotal:    ", total_reliable_pages());
 	show_val_kb(m, "ReliableUsed:     ", used_reliable_pages());
 	show_val_kb(m, "ReliableBuddyMem: ", free_reliable_pages());
+
+	if (pagecache_reliable_is_enabled()) {
+		unsigned long num = 0;
+
+		num += global_node_page_state(NR_LRU_BASE + LRU_ACTIVE_FILE);
+		num += global_node_page_state(NR_LRU_BASE + LRU_INACTIVE_FILE);
+		show_val_kb(m, "FileCache:        ", num);
+	}
 }
 
 static int __init setup_reliable_debug(char *str)
-- 
2.17.1

